### PR TITLE
add service discovery to consul service

### PR DESCRIPTION
**Pull request by ghost**
@svenwltr @das-vinculum @bjoernhaeuser added a service discovery functionality. Used random for load balancing because its easier to keep thread safe than a round robin would be.

### CODE REVIEW

File: src/main/java/com/rebuy/consul/ConsulService.java [hunk 1]
```diff
@@ -42,4 +48,17 @@ public void unregister()
         client.agentServiceDeregister(service.getId());
         logger.info("unregistered service in consul: {}", service.toString());
     }
+
+    public CatalogService findService(String serviceName)
```
Comment (svenwltr) [inline src/main/java/com/rebuy/consul/ConsulService.java]: I would like to use a name, that implies it uses round robin.

Comment (ghost) [inline src/main/java/com/rebuy/consul/ConsulService.java]: getOneService or getOneRandomService?

Comment (svenwltr) [inline src/main/java/com/rebuy/consul/ConsulService.java]: or getRandomService? Because the return type implies it is one.

**Comment by svenwltr** [comment]
:+1:

**Comment by bjoernhaeuser** [comment]
And consul itself always returns the list in the same order?

**Comment by ghost** [comment]
if I ask the catalog for the service name yes it is always the same order, consul has a query option to always return the shortest path, but I couldnt find it in our client and I also didn't look into how it calculates that distance.

**Comment by bjoernhaeuser** [comment]
distance is the one which is has the shortest/smallest network latency.

**Comment by ghost** [comment]
that much I know but I dont know how its measured

**Comment by bjoernhaeuser** [comment]
From the RAFT protocol for cluster membership

**Comment by ghost** [comment]
unfortunately our library doesn't support the "near=_agent" option in its query parameters. If we want this we could fix it and try to get it into upstream.

**Comment by bjoernhaeuser** [comment]
Okay ;)

File: src/main/java/com/rebuy/consul/ConsulService.java [hunk 2]
```diff
@@ -42,4 +48,17 @@ public void unregister()
         client.agentServiceDeregister(service.getId());
         logger.info("unregistered service in consul: {}", service.toString());
     }
+
+    public CatalogService getRandomService(String serviceName)
```
Comment (florian-zeidler) [inline src/main/java/com/rebuy/consul/ConsulService.java]: I would consider using our own wrapper, that would keep the wrapped client contained.

**Comment by florian-zeidler** [comment]
What needs to be done to get a thumb under that one?
I do not see anything.

**Comment by florian-zeidler** [comment]
@rebuy-de/it-platform Please review! Any recommendations?

File: src/main/java/com/rebuy/consul/Service.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.rebuy.consul;
+
+public class Service
+{
+    String hostname;
```
Comment (svenwltr) [inline src/main/java/com/rebuy/consul/Service.java]: maybe final?

**Comment by svenwltr** [comment]
:+1:

**Comment by bjoernhaeuser** [comment]
Should we also add a method to get all services?

Otherwise :+1:

**Comment by florian-zeidler** [comment]
I do not need all services, but I need specific services.

**Comment by bjoernhaeuser** [comment]
We are going to need that, for example for kafka/zookeeper :)

**Comment by florian-zeidler** [comment]
Yes, but later

**Pull request merged**